staging/src/k8s.io/client-go/tools/cache/controller.go

```diff
@@ -210,16 +210,70 @@ func (c *controller) processLoop() {
 //     it will get an object of type DeletedFinalStateUnknown. This can
 //     happen if the watch is closed and misses the delete event and we don't
 //     notice the deletion until the subsequent re-list.
+//
+// Implementers MUST ALSO implement OnAdd, from either
+// ResourceEventHandlerInspecific or ResourceEventHandlerDetailed (preferred).
 type ResourceEventHandler interface {
-	OnAdd(obj interface{})
 	OnUpdate(oldObj, newObj interface{})
 	OnDelete(obj interface{})
 }
 
+// ResourceEventHandlerInspecific has the (old) version of the OnAdd method
+// which doesn't propagate isInInitialList. The method is split up like this to
+// avoid breaking pre-existing callers.
+type ResourceEventHandlerInspecific interface {
+	ResourceEventHandler
+	OnAdd(obj interface{})
+}
+
+// ResourceEventHandlerDetailed has the (new) version of the OnAdd method
+// which *does* propagate isInInitialList.
+//
+// TODO: We're assuming that entries in the initial list all show up as Add
+// events. Check this assumption.
+type ResourceEventHandlerDetailed interface {
+	ResourceEventHandler
+	OnAdd(obj interface{}, isInInitialList bool)
```

Comment:
Will there be a followup just breaking the contract and using  this?  I have a personal preference for only have the API I want to support.

Reply:
I added a commit adding notices to the things that would change -- I'm not totally sure it's worth the complaints we'll get but I don't care too much